src/strands/event_loop/event_loop.py

```diff
@@ -132,14 +132,14 @@ async def event_loop_cycle(agent: "Agent", invocation_state: dict[str, Any]) ->
             model_id=model_id,
         )
         with trace_api.use_span(model_invoke_span):
-            tool_specs = agent.tool_registry.get_all_tool_specs()
-
             agent.hooks.invoke_callbacks(
                 BeforeModelInvocationEvent(
                     agent=agent,
                 )
             )
 
+            tool_specs = agent.tool_registry.get_all_tool_specs()
```

Comment:
Historical context: Originally this was moved to *before* the hook, as we were passing it in as a parameter to `BeforeModelInvocationEvent`; that was removed as part of the initial release, because ToolSpec's themselves seemed too low-level compared to exposing AgentTool.

Off the top of mind, I wonder if we should be instead passing the dictionary of tools to be included in the event, and then the hook callbacks can adjust it as needed as part of the hook execution, which might be a cheaper alternative then adding/removing tools from the agent per hook?

Reply:
It would be a slight breaking change to adjust that in the future, as we'd then be changing it so that the `BeforeModelInvocationEvent` would be driving which tools are available vs calling into the tool_registry.

That said, I think there's a path forward where it's backwards compatible (just more complicated) so I'm good with this change.